BrainPortal/app/models/tool_config.rb

```diff
@@ -326,22 +326,43 @@ def cbrain_task_class
 
   # Returns an array of full paths to the Singularity overlay files that
   # need to be mounted, as configured by the admin. Some of them might
-  # be patterns and will need to be resolved at run time.
+  # be patterns and will need to be resolved at run time. The dsl is
+  #      # A file located on the cluster
+  #        file:/path/to/file.squashfs
+  #      # All overlays configured for DP 1234
+  #         dp:1234
+  #      # CBRAIN db registered file
+  #         userfile:1234
   def singularity_overlays_full_paths
     specs = self.singularity_overlays_specs.presence
     return [] if specs.blank?
 
-    specs = specs.split(/[\s,]+/).map(&:presence).compact
-    specs.map do |spec|
-      # Full path or pattern
-      next spec if spec =~ /^\// # just use that
-      # DP spec dp:123 or dp:name
-      id_or_name = spec.sub(/^dp:/i,"")
-      dp = DataProvider.where_id_or_name(id_or_name).first
-      cb_error "Can't find DataProvider #{id_or_name} for fetching overlays" if ! dp
-      dp_ovs = dp.singularity_overlays_full_paths rescue nil
-      cb_error "DataProvider #{id_or_name} does not have any overlays configured." if dp_ovs.blank?
-      next dp_ovs
+    # break onto commands based on comas, whitespace or comments
+    specs = parsed_overlay_specs(specs)
+    specs.map do |knd, id_or_name|
+
+      # Old style file spec (legacy, to be removed)
+      next knd if knd =~ /^\//  # FIXME delete it after successful migration
+
+      case knd
+      when 'dp'
+        dp = DataProvider.where_id_or_name(id_or_name).first
+        cb_error "Can't find DataProvider #{id_or_name} for fetching overlays" if ! dp
+        dp_ovs = dp.singularity_overlays_full_paths rescue nil
+        cb_error "DataProvider #{id_or_name} does not have any overlays configured." if dp_ovs.blank?
+        dp_ovs
+      when 'file'
+        cb_error "Provide absolute filepath in spec #{spec}." if (Pathname.new id_or_name).relative?
+        id_or_name  # for local file, it could be full file name (no ids)
+      when 'userfile'
+        # db registered file, note admin can access all files
+        userfile = Userfile.where(:id => id_or_name).last
```

Comment:
Small adjustements:
1) fetch the file using 'SingleFile' instead of 'Userfile', that way we make sure the admin has configured the ID of a file, not a directory
2) make the error message more clear the problem is in the overlay spec, because it will be difficult to understand the message `Userfile 12345 not found` in the logs if we don't know this error occursed during the stage where we build the overlays

Reply:
ok